ethcore/src/evm/interpreter.rs

```diff
@@ -644,10 +640,8 @@ impl Interpreter {
 				let init_off = stack.pop_back();
 				let init_size = stack.pop_back();
 				let return_code = mem.read_slice(init_off, init_size);
-				let gas_left = try!(ext.ret(&gas, &return_code));
-				return Ok(InstructionResult::StopExecutionWithGasLeft(
-					gas_left
-				));
+
+				return Ok(InstructionResult::StopExecutionNeedsReturn(gas, return_code.to_owned()))
```

Comment:
Interesting! The reason I switched to a Vec here was to avoid borrow checker errors -- `mem` is a local variable, so we can't return a slice into it. But mem doesn't get used at all afterwards, so I think something like this would be an acceptable solution:

  - move the return code to the beginning of the memory vector
  - truncate the vector to init_size

I think this can be done with

`StopExecutionNeedsReturn(U256, Vec<u8>)` -> `StopExecutionNeedsReturn(usize, usize)`
holding `init_off` and `init_len`
```
for _ in mem.drain(..init_len) {}
for _ in mem.drain(init_size..) {}

Ok(GasLeft::NeedsReturn(gas, mem))
```

the ideal thing to do here is a `ptr::copy` and a `vec.set_len()`, but I'd prefer to avoid unsafety if possible. I think LLVM should be able to optimize at least the second call to drain, but whether it could raise the first to a `memcpy` intrinsic is another question.